renderdoc/driver/vulkan/vk_initstate.cpp

```diff
@@ -2145,15 +2145,22 @@ void WrappedVulkan::Apply_InitialState(WrappedVkRes *live, const VkInitialConten
     {
       if(it->start() >= initial.mem.size)
         continue;
+      VkDeviceSize start = it->start();
       VkDeviceSize finish = RDCMIN(it->finish(), initial.mem.size);
-      VkDeviceSize size = finish - it->start();
+      VkDeviceSize size = finish - start;
       switch(it->value())
       {
         case eInitReq_Clear:
-          ObjDisp(cmd)->CmdFillBuffer(Unwrap(cmd), Unwrap(dstBuf), it->start(), size, 0);
-          fillCount++;
-          break;
-        case eInitReq_Copy: regions.push_back({it->start(), it->start(), size}); break;
+          if((start & 3) == 0 && ((finish & 3) == 0 || finish >= initial.mem.size))
+          {
+            if(finish >= initial.mem.size)
+              size = VK_WHOLE_SIZE;
+            ObjDisp(cmd)->CmdFillBuffer(Unwrap(cmd), Unwrap(dstBuf), start, size, 0);
+            fillCount++;
+            break;
+          }
+        // else fallthrough to case eInitReq_Copy
```

Comment:
This should be correct but it seems like a big hammer to fix the problem. If there's a feasible solution where we can just do a few more bytes of copying and few less bytes of copying then that would be nicer since we'd still get a lot of the benefit. In the worst case if all the clear regions are unaligned and we end up copying everything for the sake of a few bytes.

The `Intervals<InitReqType> resetReq` is initialised above locally in this function from the resource's `Intervals<FrameRefType> rangeRefs`. Could we do something there where we align `eInitReq_Copy` conservatively outwards to the nearest 4-byte boundary, and `eInitReq_Clear` conservatively inwards? That shouldn't cause an overlap nor should it do any harm (except for maybe a few extra bytes of copy here and there that would otherwise be `eInitReq_None`.

Or alternatively maybe we could split an `eInitReq_Clear` that's unaligned into `{unaligned copy} {aligned clear} {unaligned copy}` then do a pass using `mergeLeft()` to combine any adjacent copies with the tiny unaligned copies?

There might be a good reason not to do either of these, but if not it would be a nice improvement.

Reply:
Yes, I agree with all of this. I was just wasn't sure about the added complexity to optimize a case of unknown frequency, particularly given that I don't have an easy way to test the change.

I think the simplest, most correct case is to just align all the copy/clear intervals out by 4 when constructing resetReq. Since we are using the interval update function is the max reset requirement, this should correctly make any overlap between a copy and a clear into a copy. This should also correctly handle the case when a clear interval is isolated--it should be expanded outward in this case. I will update the PR.